Character rework

atomic-junky opened the pull request:
I'm trying to rework the characters by adding a Character node and completely overhauling the way they're defined, adding lots of features like portraits and animations.
For the moment I've made a mockup of the character editor and the field in the root node as well as the new CharacterNode works.

<img src="https://github.com/user-attachments/assets/99cdb736-defd-49c4-be66-a3d2b3efdabf" width="500">

> The NodeCharacter and the SidePanel


<img src="https://github.com/user-attachments/assets/c91b39fd-80a4-4b45-bc94-b2d67d691e2d" width="300">

> The new character fields defines inside the NodeRoot

<img src="https://github.com/user-attachments/assets/8c6bd10b-90bc-4d52-ac9c-ca7f6f945072" width="500">

> The default character edit (it opens when you click on the "Edit" button)

<img src="https://github.com/user-attachments/assets/4f6c1fd2-10a8-4a8a-9953-a0f0a66b8844" width="500">

> The character edit with the portrait type set to Animation

I honnestly need help so I don't mess it up. The new CharacterFields are a bit hardcoded and perhaps work a bit strangely. But more importantly, for the fields in the character edit, they're not MonologueFields and I have no idea how to do this properly. 
*However, if you have any other ideas or a different vision of how it should be done, I'm really interested.*

*I hope you're well and I wish you happy holidays! (reply to me when you have the time and mood)*

atomic-junky committed:
Add NodeCharacter

atomic-junky committed:
Merge branch 'dev' into feature/node-character

atomic-junky committed:
Fix empty collapsible field

atomic-junky committed:
Fix CharacterNode preview

atomic-junky committed:
Update sentence_node.gd

atomic-junky committed:
Fix hidden collapsible field

atomic-junky committed:
Support collapsible field in cascade

atomic-junky committed:
Update character_node.gd

atomic-junky committed:
Rework MonologueCharacter

atomic-junky committed:
The Narrator is now a protected character

atomic-junky committed:
CharacterEdit mockup

atomic-junky committed:
Portraits section + other things

atomic-junky committed:
Selectable portrait option

atomic-junky committed:
Update timeline

atomic-junky committed:
New buttons to add cell and layer

atomic-junky committed:
Register portrait list

atomic-junky committed:
Register default portrait

atomic-junky committed:
Rework portrait option

atomic-junky committed:
Register Portrait Settings

atomic-junky committed:
Fix wrong character edit section visible when open

atomic-junky committed:
Timeline cells are now moveable

RailKill commented:
happy new year! i'm in a game jam now, will probably only be able to look at it in 1-2 weeks time. otherwise, if you feel it's good enough feel free to merge 😄

atomic-junky committed:
Register Animation

atomic-junky committed:
Preview for Image is now working

atomic-junky commented:
No problem, happy jam!
*I think it's going to take a while before it works.*

atomic-junky committed:
Fix some mistakes

atomic-junky committed:
Support mirror for the preview

atomic-junky committed:
Fix changes not being saved in some cases

atomic-junky committed:
Update preview when switching between characters

atomic-junky committed:
Better cell selection + image path to cell

atomic-junky committed:
preview timeline cell

atomic-junky committed:
Start working on animation preview

atomic-junky committed:
Support multiple layers for the preview

atomic-junky committed:
Update preview when switching between portrait types

RailKill committed:
Fix bugs and debugger warnings

RailKill committed:
Set default character index to 0

atomic-junky committed:
Character edit main section now use properties

RailKill committed:
Merge branch 'feature/node-character' of https://github.com/atomic-junky/Monologue into feature/node-character

RailKill committed:
Persist character edit data

RailKill committed:
Fix anchor preset

RailKill committed:
Fix character edit index

RailKill committed:
Move character edit data to Character dictionary

RailKill committed:
Allow reload when undo/redo different character index

RailKill commented:
i added CharacterHistory which extends PropertyHistory so that the character edit data can work in undo/redo and be able to swap between characters. this class will also update the character data in its corresponding graph_edit via `set_property()` and thus will persist in the .json as well.

i think this should be a good enough foundation. as long as your ui extends CharacterEditSection, and they are known to `character_edit.gd::trickle()` and `character_edit.gd::_to_dict()`, any changes to its variables of type Property should be automatically picked up

atomic-junky committed:
Portrait list use MonologueList + PortraitHistory

atomic-junky committed:
Portrait list support scroll

atomic-junky committed:
Improve portrait list

atomic-junky committed:
Revert changes in monologue_list.gd

atomic-junky committed:
Update gdUnit4 to 4.5

atomic-junky committed:
Improve PortraitOption

atomic-junky committed:
Minor changes on PortraitOption

atomic-junky commented:
Your changes work really well, thank you very much. But I tried to make the portrait list work but the “portraits” property doesn't retain changes. 
If you close character edit and then open it, the changes are not saved because the value of the “portraits” property is empty. I have no idea why.
If you can figure it out, that would be amazing.

What's left to do is to make the portrait settings section work, which should properly show the settings of the selected portrait when switching between them. And the Timeline will have to work too, probably by making it a field.

RailKill committed:
Merge branch 'dev' into feature/node-character

RailKill committed:
Fix portrait list not picking up changes

RailKill committed:
Load portrait list

RailKill committed:
Track default portrait change

RailKill committed:
Remove unused broken portrait_section_container.gd

RailKill commented:
fixed the portrait list. add/delete and set default changes now track and persist. the issue was that the property `change` signal was not connected. in `MonologueGraphNode` this was done in `_ready()`, so i added code in `CharacterEditSection::_ready()` to connect the signal.

things still not tracked yet:
- renaming portrait option in the portrait list
- portrait settings
- timeline

great work on the UI btw, seems like it'll be very cool to animate character portraits. good progress. i'll take a break for now, will come back to this next time 😄

atomic-junky commented:
I tried to get the Portrait settings section to work, but I couldn't get it to work. My problem is that as soon as a portrait is selected, the field values have to change. But do the fields have to be in the portrait settings section or in the AbstractPortraitOption class? And this problem also applies to the timeline section.

atomic-junky committed:
Merge branch 'dev' into feature/node-character

atomic-junky committed:
Godot 4.4 stuff

RailKill committed:
Merge branch 'dev' into feature/node-character

RailKill committed:
Ensure default portrait value gets loaded

RailKill committed:
Implement portrait settings history

RailKill commented:
latest commit now updates some of the portrait settings. for `CharacterEditSection` you can call `_from_dict()` to load new data, it will update the fields via `Property::show()` which accepts `panel` as an argument but that `panel` can be any `Control`, so the fields can actually be created anywhere so as long as they are tied to a property. if you want to refresh fields that are already present, you can use `Property::propagate()` or just recreate the field

atomic-junky commented:
it works pretty well, but when you add a portrait, it resets all the portraits

atomic-junky committed:
Merge branch 'dev' into feature/node-character

atomic-junky committed:
Fix portrait settings data

atomic-junky committed:
Connect portrait name

atomic-junky committed:
Switch timeline visibility

atomic-junky commented:
The portrait settings now work (normally). It's just the portrait name that doesn't work.

atomic-junky committed:
Zoomable timeline

atomic-junky committed:
Sub selection + better zoom

atomic-junky committed:
Character Edit image preview

atomic-junky committed:
Timeline cell rework

atomic-junky committed:
Better timeline layer

atomic-junky committed:
Various tweaks for animation layers

atomic-junky commented:
I've made a lot of changes to the animation timeline. What's left to do would be to be able to load an image or a spritesheet to a layer (that would be pretty cool), make the preview work with the animation (there's a few bricks of code left for that) and save the changes with properties (but I'm not sure how to do that). After that, I think this pr will be ready for review and I think Monologue would have all the necessary features.

By the way, I'd love to hear your opinion on how the new version of the timeline works, if it's intuitive and handy. Thank you again for all the contributions you brought to the project.

atomic-junky committed:
timeline layer to spritesheet

atomic-junky committed:
Fix cell texture not loading

atomic-junky committed:
Update node character

atomic-junky committed:
Update node character in the node picker

atomic-junky committed:
Start trying to make Timeline a field (it's broken)

atomic-junky committed:
Save/load animation

atomic-junky committed:
Fix weird exposure issue

atomic-junky committed:
fix variable shadowing

atomic-junky committed:
Deletable layers

atomic-junky committed:
Fps value

atomic-junky committed:
wip threaded image loader

atomic-junky committed:
Better preview + multiple image selection

atomic-junky committed:
Better cell selection

atomic-junky committed:
Fix some bugs

atomic-junky committed:
One shot portrait settings property

There's a weird bug with this property visibility

atomic-junky committed:
Keep frame exposure and position when import multiple frames

atomic-junky committed:
Renamable portraits

atomic-junky committed:
Various tweaks

atomic-junky committed:
Remove eye button (layer visibility)

It's useless

atomic-junky committed:
Delete ReorderableContainer addons

atomic-junky committed:
Fix CharacterNode portrait and remove display_variant

- Rename speakers variable to characters inside the graph_edit
- Fix CharacterNode portrait field
- Remove display_variant field who is useless thanks to the CharacterNode

atomic-junky committed:
Fix warnings

atomic-junky committed:
Fix test_sentence_node.gd not working

atomic-junky committed:
Remove a useless button

atomic-junky commented:
I believe this PR is feature‑complete. It still needs extensive testing, but there’s a known issue: the **one\_shot** field in the character settings doesn’t display correctly, and I’m not sure why.

Below is a list of all the changes in this PR:

* **Character Management**

  * Characters are now edited in a new **CharacterEdit** panel.
  * The special character `_NARRATOR` is locked—cannot be deleted or renamed.
  * Each character can have a **nickname** (useful for recognizing its name in a sentence).
  * Each character has a **default display name** in addition to its internal name.
  * Characters have a **description** field.

* **Portraits**

  * Characters can have multiple portraits, and one must be marked as the default.
  * Portraits are edited inside the **CharacterEdit** panel.
  * Each portrait has a **type**: either **Image** or **Animation**.
  * If the portrait type is **Image**, there’s an **image\_path** field.
  * There’s a new **Vector** field (with `x` and `y` spinboxes) used for offsets.
  * Portraits have an **offset** property (a Vector) and a **mirror** (toggle) property.
  * If the portrait type is **Animation**, it has a **one\_shot** (toggle) property.
  * Animated portraits also have an **animation** property (a Timeline).

* **Timeline (for animated portraits)**

  * A new **Timeline** field appears under the Timeline section of the CharacterEdit panel.
  * Timelines have an **FPS** spinbox.
  * Timelines contain **layers**, and layers can be either “layer groups” or “cell layers.”
  * **Cell layers** can hold individual frames (“cells”).
  * When hovering over a cell, buttons appear to add or remove exposure.
  * Selecting a cell lets you load one or multiple images at once.
  * If you load multiple images, new cells are created immediately after the selected one, each inheriting that cell’s exposure setting.

* **Preview**

  * Both Image‑type and Animation‑type portraits can be previewed in the Preview section.

* **SentenceNode Changes**

  * The **display\_variant** property has been removed from SentenceNode.

* **New CharacterNode**

  * There’s a new **CharacterNode** type.
  * CharacterNode has two properties: **character** and **action\_type**.
  * The `action_type` can be **Join**, **Leave**, or **Update**.
  * Depending on the action type, different properties are shown or hidden: * **\_position**, **join\_animation**, **leave\_animation**, **update\_animation**, **portrait**, **duration**, **\_z\_index**, **mirrored**.
  * The available options for **join\_animation** and **leave\_animation** are: * Default, None, Fade In, Slide In Auto, Slide In Down, Slide In Left, Slide In Right, Slide In Up.
  * The available options for **update\_animation** are: * Default, None, Bounce, Shake.

There may be a few minor additions not listed here, but this covers the most important changes.

atomic-junky committed:
Fix renaming a character reset it

When the value of a property of type Dictionary is changed, it will be merged with the previous value to avoid losing any data.

atomic-junky committed:
Reset selected portrait index when opening CharacterEdit

atomic-junky committed:
Fix preview

atomic-junky committed:
Fix dictionary value property merging

atomic-junky committed:
Fix CharacterNode Portrait field

atomic-junky committed:
Merge branch 'dev' into feature/node-character

atomic-junky committed:
Resolve conflicts (and more)

atomic-junky committed:
Fix + improvements

atomic-junky committed:
Better portrait option selection

atomic-junky committed:
Fix failure of GdUnit tests

atomic-junky committed:
Create test-export.yml

atomic-junky committed:
Update test.yml

atomic-junky committed:
Replace godot_executable_download_url and godot_export_templates_download_url

atomic-junky committed:
Update export presets

atomic-junky committed:
Fix Character Edit things

atomic-junky committed:
Update export_presets.cfg

atomic-junky committed:
Fix export workflows

atomic-junky committed:
Bump GDUnit4 version to 5

atomic-junky committed:
A lot of qol + other things (fix and more)

atomic-junky committed:
update github workflows ?

atomic-junky committed:
???

atomic-junky committed:
Forget about this

atomic-junky committed:
Run Gdformat from godot-gdscript-toolkit

atomic-junky committed:
Delete .pre-commit-config.yaml

atomic-junky committed:
I'm so stupid

atomic-junky committed:
Fix GraphNode not centered when added to GraphEdit

atomic-junky committed:
Improve timeline ui

atomic-junky committed:
NodeBackground now use ImageLoader

atomic-junky committed:
test thing

atomic-junky committed:
Update EndPathNode

atomic-junky committed:
Fix EndPathNode

atomic-junky committed:
Settings button

atomic-junky committed:
WelcomeWindow can be exited

atomic-junky committed:
Center graph node picker when summon with a button

atomic-junky committed:
GDUnit things

atomic-junky committed:
Searchbar mockup
